src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc

```diff
@@ -177,6 +178,96 @@ VKAPI_ATTR void VKAPI_CALL vkCmdDrawIndexed(
 
 // Other intercepted vulkan functions.
 
+//
+// Our vkCreateGraphicsPipelines function.
+//
+VKAPI_ATTR VkResult VKAPI_CALL vkCreateGraphicsPipelines(
+    VkDevice device, VkPipelineCache pipelineCache, uint32_t createInfoCount,
+    const VkGraphicsPipelineCreateInfo* pCreateInfos,
+    const VkAllocationCallbacks* pAllocator, VkPipeline* pPipelines) {
+  GlobalData* global_data = GetGlobalData();
+  DeviceData* device_data =
+      global_data->device_map.Get(DeviceKey(device))->get();
+
+  // Call the original function.
+  VkResult result = device_data->vkCreateGraphicsPipelines(
+      device, pipelineCache, createInfoCount, pCreateInfos, pAllocator,
+      pPipelines);
+
+  if (result == VK_SUCCESS) {
+    // Deep copy all create info structs.
+    for (uint32_t i = 0; i < createInfoCount; i++) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      VkGraphicsPipelineCreateInfo create_info = DeepCopy(pCreateInfos[i]);
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      device_data->graphics_pipelines.Put(pPipelines[i], create_info);
+
+      // Register the pipeline as a user of shader module of each stage so the
+      // shader modules can be freed from the memory when no longer used by any
+      // pipeline.
```

Comment:
Some alternatives:

1. We could simply add a copy of the `ShaderModuleData` into the `GraphicsPipelineData` object (assuming you add `GraphicsPipelineData` as I suggested). But perhaps games are likely to create many pipelines with the same shaders (and keep them around), which could take up large amounts of memory.
2. You could just increment a reference count on the ShaderModuleData, instead of storing the list of pipelines. Then, when a pipeline is destroyed, you can iterate over the shader modules in the `VkGraphicsPipelineCreateInfo` and deref them. 

Actually, there is another problem with the current solution and solution 2. A game creates a shader module, uses it to create a pipeline, destroys the shader module, creates a new shader module. The new shader module might get the same handle as the destroyed handle. Thus, I suggest you try solution 1 for now.

Reply:
I fixed this with solution 1, but I used shared pointers to avoid copying shader module data.

Reply:
Nice!